cmake/ranges_flags.cmake

```diff
@@ -0,0 +1,184 @@
+# Copyright Louis Dionne 2015
+# Copyright Gonzalo Brito Gadeschi 2015
+# Distributed under the Boost Software License, Version 1.0.
+# (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+#
+# Setup compiler flags (more can be set on a per-target basis or in
+# subdirectories)
+
+# Compilation flags
+include(CheckCXXCompilerFlag)
+macro(ranges_append_flag testname flag ${ARGN})
+    check_cxx_compiler_flag("${flag} ${ARGN}" ${testname})
+    if (${testname})
+        add_compile_options(${flag} ${ARGN})
+    endif()
+endmacro()
+
+# All compilation flags
+# Language flag: version of the C++ standard to use
+message("[range-v3]: C++ std=${RANGES_CXX_STD}")
+ranges_append_flag(RANGES_HAS_CXXSTD "-std=c++${RANGES_CXX_STD}")
+
+# Enable all warnings and make them errors:
+ranges_append_flag(RANGES_HAS_WERROR -Werror)
+ranges_append_flag(RANGES_HAS_WX -WX)
+ranges_append_flag(RANGES_HAS_WALL -Wall)
+ranges_append_flag(RANGES_HAS_WALL -Wextra)
+ranges_append_flag(RANGES_HAS_WEVERYTHING -Weverything)
+ranges_append_flag(RANGES_HAS_PEDANTIC -pedantic)
+ranges_append_flag(RANGES_HAS_PEDANTIC_ERRORS -pedantic-errors)
+
+# Selectively disable those warnings that are not worth it:
+ranges_append_flag(RANGES_HAS_WNO_CXX98_COMPAT -Wno-c++98-compat)
+ranges_append_flag(RANGES_HAS_WNO_CXX98_COMPAT_PEDANTIC -Wno-c++98-compat-pedantic)
+ranges_append_flag(RANGES_HAS_WNO_WEAK_VTABLES -Wno-weak-vtables)
+ranges_append_flag(RANGES_HAS_WNO_PADDED -Wno-padded)
+ranges_append_flag(RANGES_HAS_WNO_MISSING_PROTOTYPES -Wno-missing-prototypes)
+ranges_append_flag(RANGES_HAS_WNO_MISSING_VARIABLE_DECLARATIONS -Wno-missing-variable-declarations)
+ranges_append_flag(RANGES_HAS_WNO_DOCUMENTATION -Wno-documentation)
+ranges_append_flag(RANGES_HAS_WNO_DOCUMENTATION_UNKNOWN_COMMAND -Wno-documentation-unknown-command)
+ranges_append_flag(RANGES_HAS_WNO_OLD_STYLE_CAST -Wno-old-style-cast)
+ranges_append_flag(RANGES_HAS_WNO_OLD_STYLE_CAST -Wno-shadow)
+
+if(RANGES_ENV_MACOSX)
+  ranges_append_flag(RANGES_HAS_WNO_GLOBAL_CONSTRUCTORS -Wno-global-constructors)
+  ranges_append_flag(RANGES_HAS_WNO_EXIT_TIME_DESTRUCTORS -Wno-exit-time-destructors)
+endif()
+
+if(RANGES_CXX_COMPILER_GCC AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS "5.0"))
+  ranges_append_flag(RANGES_HAS_WNO_MISSING_FIELD_INITIALIZERS -Wno-missing-field-initializers)
+endif()
+
+
+# Workaround missing __extern_always_inline in clang under linux
+if(RANGES_ENV_LINUX AND RANGES_CXX_COMPILER_CLANG)
+  ranges_append_flag(RANGES_HAS___EXTERN_ALWAYS_INLINE -D__extern_always_inline=inline)
```

Comment:
That sounds familiar. It would be really nice to document the newest Clang version that triggers this, because I think it's getting close to being no longer relevant.